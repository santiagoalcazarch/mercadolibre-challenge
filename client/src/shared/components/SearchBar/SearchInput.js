
import React from 'react'
import PropTypes from 'prop-types'
import * as Yup from 'yup';
import { useFormik } from 'formik';
import { useDispatch } from 'react-redux';
import { useNavigate } from "react-router-dom";

import SearchImg from '../../assets/icons/ic_Search@2x.png'
import { actions } from '../../../ecommerce/redux/items/actions';
import { LIST_ITEM } from '../../../routes/routes';


const SearchInput = ({ serachInitValue = "" }) => {

  const dispatch = useDispatch( );
  const navigate = useNavigate();

  const formik = useFormik({
    initialValues: { search: serachInitValue },
    enableReinitialize: true,
    validationSchema: Yup.object().shape({
      search: Yup.string()
        .max(100, '¡Ingresaste demasidos valores!')
        .required('¿Estas buscando algo? ¡Ingresalo!'),
    }),
    onSubmit: values => {
      dispatch( actions.searchAndSetItemsList( values.search ) );
      navigate(`${LIST_ITEM}?search=${values.search}`);
    },
  });

  return (
    <form onSubmit={formik.handleSubmit}>
      <div className="input-group my-3 search-input ml-4">
        <input 
          type="text"
          id="search"
          name="search"
          onChange={formik.handleChange}
          value={formik.values.search ?? ""}
          className="form-control" 
          placeholder="Nunca dejes de buscar" 
          aria-label="serach-action"
          aria-labelledby='serach-action'
        />
        <span className="input-group-text m-0 p-0" id="serach-action">
          <button 
            className="search-btn" 
            id="search-btn"
            type="submit"
          >
            <img src={SearchImg} alt="Search" />
          </button>
        </span>
      </div>
      { formik.errors.search && formik.touched.search && 
        <div className="ml-4 mb-2">
          <span className="invalid-msg">{formik.errors.search}</span>
        </div>
      }
    </form>
      
  )
}

SearchInput.propTypes = {
  serachInitValue: PropTypes.string
}

export default SearchInput