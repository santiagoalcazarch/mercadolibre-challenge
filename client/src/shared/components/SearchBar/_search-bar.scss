
.search-layout{

    background-color: $mainYellow;

    .search-input{
        font-size: 18px;
        color: $mainBlack;

        .form-control:focus {
            box-shadow: none;
        }
        
        input{
            border: none;
            
        }

        .search-btn{
            background-color: transparent;
            border: none;
            padding: 0.5rem 0.8rem;

            img{
                max-width: 18px;
            }
        }
    }

    .search-icon{
        background-color: $mainGray;
    }

}