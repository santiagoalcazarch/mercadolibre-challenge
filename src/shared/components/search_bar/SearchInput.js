
import React from 'react'
import PropTypes from 'prop-types'
import SearchImg from '../../assets/icons/ic_Search@2x.png'

const SearchInput = props => {
  return (
    <div className="input-group my-3 search-input mx-4">
      <input 
        type="text" 
        className="form-control" 
        placeholder="Nunca dejes de buscar" 
        aria-label="serach-action"
        aria-labelledby='serach-action'
      />
      <span className="input-group-text m-0 p-0" id="serach-action">
        <button className="search-btn">
          <img src={SearchImg} alt="Search" />
        </button>
      </span>
    </div>
  )
}

SearchInput.propTypes = {}

export default SearchInput