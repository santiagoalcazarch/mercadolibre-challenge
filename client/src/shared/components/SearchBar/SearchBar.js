import React, { useEffect, useState } from 'react'
import { useDispatch } from 'react-redux';

import SearchInput from './SearchInput'
import Logo from '../Logo/Logo'
import { actions } from '../../../ecommerce/redux/items/actions';
import { useQuery } from '../../../helpers/hooks/useQuery'

const SearchBar = () => {

  let query = useQuery();
  const dispatch = useDispatch();
  const [serachInitValue, setSerachInitValue] = useState('');

  useEffect(() => {
    const searchParam = query.get('search');
    if ( searchParam !== undefined ) {
      setSerachInitValue(searchParam);
      dispatch( actions.searchAndSetItemsList( searchParam ) );
    }
  }, [query, dispatch])

  return (
    <div className="container-fluid search-layout">
      <div className="container d-flex align-items-center">
        <Logo />
        <SearchInput serachInitValue={serachInitValue} />
      </div>
    </div>
  )
}

export default SearchBar