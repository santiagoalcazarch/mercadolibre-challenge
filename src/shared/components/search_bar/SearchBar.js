import React from 'react'
import PropTypes from 'prop-types'
import Logo from '../logo/Logo'
import SearchInput from './SearchInput'

const SearchBar = props => {
  return (
    <div className="container-fluid search-layout">
      <div className="container d-flex align-items-center">
        <div>
          <Logo />
        </div>
        <SearchInput />
      </div>
    </div>
  )
}

SearchBar.propTypes = {}

export default SearchBar