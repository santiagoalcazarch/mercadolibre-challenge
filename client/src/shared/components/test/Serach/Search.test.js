import { mount } from "enzyme"
import { MemoryRouter } from "react-router-dom";
import { useDispatch } from 'react-redux';

import SearchInput from "../../SearchBar/SearchInput"
import SearchBar from "../../SearchBar/SearchBar";

jest.mock("react-redux");

describe('Search-bar test', () => {

  const initSearch = "iphone";

  test('Debería hacer match con el snapshot', () => {
    const wrapper = mount( 
      <MemoryRouter> 
        <SearchInput />
      </MemoryRouter>
    )
    expect(wrapper).toMatchSnapshot();
  })

  test('Debería tener como valor inicial el paramentro enviado', () => {
    const wrapper = mount( 
      <MemoryRouter> 
        <SearchInput serachInitValue={initSearch}/>
      </MemoryRouter>
    )
    const inputValue = wrapper.find('input').prop("value");
    expect(inputValue).toBe(initSearch)
  });

  test('Debería tener como valor inicial el paramentro desde la ruta', () => {

    const dispatchJ = jest.fn();
    useDispatch.mockReturnValue(dispatchJ)

    const wrapper = mount( 
      <MemoryRouter initialEntries={["/items?search=impresora"]}>
        <SearchBar />  
      </MemoryRouter>
    )
    const inputValue = wrapper.find('input').prop("value");
    
    expect(inputValue).toBe("impresora");
    expect(dispatchJ).toHaveBeenCalled();
  });

  test('Debería no tener ningun valor inicial si no esta en la ruta de "itemsList"', () => {

    const dispatchJ = jest.fn();
    useDispatch.mockReturnValue(dispatchJ)

    const wrapper = mount( 
      <MemoryRouter initialEntries={["/item?search=impresora"]}>
        <SearchBar />  
      </MemoryRouter>
    )
    const inputValue = wrapper.find('input').prop("value");
    
    expect(inputValue).toBe("");
    expect(dispatchJ).not.toHaveBeenCalled();
  });
})
