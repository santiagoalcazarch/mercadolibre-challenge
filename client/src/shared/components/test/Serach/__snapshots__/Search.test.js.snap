// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search-bar test Deber√≠a hacer match con el snapshot 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="input-group my-3 search-input ml-4"
  >
    <input
      aria-label="serach-action"
      aria-labelledby="serach-action"
      className="form-control"
      id="search"
      name="search"
      onChange={[Function]}
      placeholder="Nunca dejes de buscar"
      type="text"
      value=""
    />
    <span
      className="input-group-text m-0 p-0"
      id="serach-action"
    >
      <button
        className="search-btn"
        id="search-btn"
        type="submit"
      >
        <img
          alt="Search"
          src="ic_Search@2x.png"
        />
      </button>
    </span>
  </div>
</form>
`;
